runtime/contract_update_validation.go

```diff
@@ -80,39 +122,18 @@ func (validator *ContractUpdateValidator) checkDeclarationUpdatability(
 		validator.currentDecl = parentDecl
 	}()
 
-	// If the same decl is already visited, then do not check again.
-	// This also avoids getting stuck on circular dependencies between composite decls.
-	if validator.visited[newDeclaration] {
-		return nil
-	}
-	validator.visited[newDeclaration] = true
```

Comment:
Is there a particular reason the recursion check was removed?

Reply:
With the most recent refactoring, the references to composite types (nominal types) are visited in a breadth-first manner, which makes this recursion check obsolete. 
i.e: when a nominal type that refers to a composite declaration is reached, the nominal type is checked only by its name and the  referred declaration is visited separately.


There is a test case: https://github.com/onflow/cadence/pull/597/files#diff-d0b635b91a31d5ad6587b489f1f630a1e317f9bb0de7d94eff8a0b4fd17847fcR301-R333 to verify such circular references.